engine/pong.h

```diff
@@ -40,15 +53,27 @@ class Pong
 
     static GameObject* FindGameObjectByName(const std::string& name);
 
-    static Text* AddText(const std::string& text, const std::string& fontPath,
-                         float scale, int pixelLineHeight = 128,
-                         const glm::vec3& position = glm::vec3(0.0f));
+    template<typename UIType, typename... Args>
+    static UIType* AddUIElement(Args&&... args)
+    {
+        static_assert(std::is_base_of<UIElement, UIType>::value, "UIType must derive from a UIElement");
```

Comment:
Ok, perfect opportunity for something fun here. 

The pattern you should use is called "SFINAE" which translates to "Substitution Failure is Not An Error". 

Without going into too much detail, in C++ before 20, there is a way to set bounds to types in a way that is like this, but provides much nicer compiler errors naturally, you may as well write one so you get use to it :) 

In fact, CppCoreGuidelines says you should ALWAYS specify bounds for every argument in a template using SFINAE in pre C++20, or concepts afterwards.


The way you would do this for this function using SFINAE is:

```
template<typename UIType, typename = std::enable_if_t<std::is_base_of_v<UIElement, UIType>>, typename... Args>
static UIType* AddUIElement(Args&&... args)
{
    // Function implementation
}
```

This will prevent any compilation if there it does NOT match the clause in the `enable_if`, and present you with a helpful diagnostic.

In C++20, we have concepts, which is the slightly (??) more sane way of doing it, it would look something like this:

```
template<typename T>
concept IsDerivedFromUIElement = std::is_base_of<UIElement, T>::value;

template<IsDerivedFromUIElement UIType, typename... Args>
static UIType* AddUIElement(Args&&... args)
{
    // Function implementation
}
```

Same exact result, you restrict the template type to conforming to some bounds.

Reply:
Aha that is awesome. Not only is it a better compiler error, but my editor can actually detect it before being compiled!